frame/staking/src/lib.rs

```diff
@@ -1472,18 +1723,179 @@ decl_module! {
+		///
+		/// 1. `min { support.total }` for each support of a winner. This value should be maximized.
+		/// 2. `sum { support.total }` for each support of a winner. This value should be minimized.
+		/// 3. `sum { support.total^2 }` for each support of a winner. This value should be
+		///    minimized (to ensure less variance)
+		///
+		/// # <weight>
+		/// E: number of edges. m: size of winner committee. n: number of nominators. d: edge degree
+		/// (16 for now) v: number of on-chain validator candidates.
+		///
+		/// NOTE: given a solution which is reduced, we can enable a new check the ensure `|E| < n +
+		/// m`. We don't do this _yet_, but our offchain worker code executes it nonetheless.
+		///
+		/// major steps (all done in `check_and_replace_solution`):
+		///
+		/// - Storage: O(1) read `ElectionStatus`.
+		/// - Storage: O(1) read `PhragmenScore`.
+		/// - Storage: O(1) read `ValidatorCount`.
+		/// - Storage: O(1) length read from `SnapshotValidators`.
+		///
+		/// - Storage: O(v) reads of `AccountId` to fetch `snapshot_validators`.
+		/// - Memory: O(m) iterations to map winner index to validator id.
+		/// - Storage: O(n) reads `AccountId` to fetch `snapshot_nominators`.
+		/// - Memory: O(n + m) reads to map index to `AccountId` for un-compact.
+		///
+		/// - Storage: O(e) accountid reads from `Nomination` to read correct nominations.
+		/// - Storage: O(e) calls into `slashable_balance_of_extended` to convert ratio to staked.
+		///
+		/// - Memory: build_support_map. O(e).
+		/// - Memory: evaluate_support: O(E).
+		///
+		/// - Storage: O(e) writes to `QueuedElected`.
+		/// - Storage: O(1) write to `QueuedScore`
+		///
+		/// The weight of this call is 1/10th of the blocks total weight.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000_000)]
+		pub fn submit_election_solution(
+			origin,
+			winners: Vec<ValidatorIndex>,
+			compact_assignments: CompactAssignments,
+			score: PhragmenScore,
+			era: EraIndex,
+		) {
+			let _who = ensure_signed(origin)?;
+			Self::check_and_replace_solution(
+				winners,
+				compact_assignments,
+				ElectionCompute::Signed,
+				score,
+				era,
+			)?
+		}
+
+		/// Unsigned version of `submit_election_solution`.
+		///
+		/// Note that this must pass the [`ValidateUnsigned`] check which only allows transactions
+		/// from the local node to be included. In other words, only the block author can include a
+		/// transaction in the block.
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000_000)]
+		pub fn submit_election_solution_unsigned(
+			origin,
+			winners: Vec<ValidatorIndex>,
+			compact_assignments: CompactAssignments,
+			score: PhragmenScore,
+			era: EraIndex,
+		) {
+			ensure_none(origin)?;
+			Self::check_and_replace_solution(
+				winners,
+				compact_assignments,
+				ElectionCompute::Unsigned,
+				score,
+				era,
+			)?
+			// TODO: instead of returning an error, panic. This makes the entire produced block
```

Comment:
@gavofyork this is the only remaining part that I told you about. I propose adding the panic/slash in a second phase. The code change would be simple and we merely panic here if `check_and_replace_solution` was  anything other than `Ok(())`.

Given
- Validators wanting to behave correctly,
- Our offchain worker code working correctly,

`submit_election_solution_unsigned` will only ever be called if a solution is an improvement to what we already have on chain (because we check the score in `ValidateUnsigned`), hence it must never panic unless if a validator attempted to put a non-improving solution on chain.